app/master/build.py

```diff
@@ -103,18 +105,30 @@ def generate_project_type(self):
         if self._project_type is None:
             raise BuildProjectError('Build failed due to an invalid project type.')
 
-    def prepare(self, subjobs, job_config):
+    def prepare(self, subjob_calculator):
         """
-        :type subjobs: list[Subjob]
-        :type job_config: JobConfig
+        :param subjob_calculator: Used after project fetch to atomize and group subjobs for this build
+        :type subjob_calculator: SubjobCalculator
         """
-        if self._project_type is None:
-            raise RuntimeError('prepare() was called before generate_project_type() on build {}.'
-                               .format(self._build_id))
+        if not isinstance(self.build_request, BuildRequest):
+            raise RuntimeError('Build {} has no associated request object.'.format(self._build_id))
+
+        if not isinstance(self.project_type, ProjectType):
+            raise RuntimeError('Build {} has no project set.'.format(self._build_id))
 
         if not self._preparation_coin.spend():
             raise RuntimeError('prepare() was called more than once on build {}.'.format(self._build_id))
 
+        self._logger.info('Fetching project for build {}.', self._build_id)
+        self.project_type.fetch_project()
+        self._logger.info('Successfully fetched project for build {}.', self._build_id)
```

Comment:
The main motivation for this change is when adding the state machine to include a "preparing" state. To enable that, the `self.project_type.fetch_project()` call has to be done within the Build class.

Also, it makes more sense to me to include all of the logic for build preparation in one place instead of doing half in BuildRequestHandler and half in Build.